libpod/container_api.go

```diff
@@ -328,44 +280,56 @@ func (c *Container) Exec(tty, privileged bool, env, cmd []string, user, workDir
 	session := new(ExecSession)
 	session.ID = sessionID
 	session.Command = cmd
-	session.PID = int(pid)
+	session.PID = pid
 	c.state.ExecSessions[sessionID] = session
 	if err := c.save(); err != nil {
 		// Now we have a PID but we can't save it in the DB
 		// TODO handle this better
-		return errors.Wrapf(err, "error saving exec sessions %s for container %s", sessionID, c.ID())
+		return defaultExecExitCode, errors.Wrapf(err, "error saving exec sessions %s for container %s", sessionID, c.ID())
 	}
 	c.newContainerEvent(events.Exec)
 	logrus.Debugf("Successfully started exec session %s in container %s", sessionID, c.ID())
 
 	// Unlock so other processes can use the container
 	if !c.batched {
 		c.lock.Unlock()
-		locked = false
 	}
 
-	var waitErr error
-	if !exited {
-		waitErr = <-chWait
+	attachErr := <-attachChan
+	if attachErr != nil {
+		logrus.Debugf(attachErr.Error())
+	}
+	close(attachChan)
+
+	exitCode, err := c.readExecExitCode(sessionID)
+	if err != nil {
+		logrus.Debugf("reading exec exit file returned error %s", err.Error())
+		if attachErr == nil {
+			attachErr = err
+		}
 	}
 
 	// Lock again
 	if !c.batched {
-		locked = true
 		c.lock.Lock()
 	}
 
 	// Sync the container again to pick up changes in state
 	if err := c.syncContainer(); err != nil {
-		return errors.Wrapf(err, "error syncing container %s state to remove exec session %s", c.ID(), sessionID)
+		return defaultExecExitCode, errors.Wrapf(err, "error syncing container %s state to remove exec session %s", c.ID(), sessionID)
 	}
 
 	// Remove the exec session from state
 	delete(c.state.ExecSessions, sessionID)
 	if err := c.save(); err != nil {
 		logrus.Errorf("Error removing exec session %s from container %s state: %v", sessionID, c.ID(), err)
 	}
-	return waitErr
+
+	// cleanup exec bundle
+	if err := c.cleanupExecBundle(sessionID); err != nil {
```

Comment:
Said this above, but we should probably just `defer` this above, right after we make the bundle